Fix workflow loading from PNG images with both workflow and parameter…

r-vage opened the pull request:
…s metadata

- Reorder handleFile() to check workflow before parameters
- Add validation to prevent JSON parse errors from crashing imports
- Fix loadGraphData() to use explicit type validation instead of falsy check
- Ensures ComfyUI-generated PNGs with both metadata types load the workflow, not parameters

Fixes issue where large workflows (e.g., 634 nodes) were replaced with basic A1111 format when importing PNG files.

## Summary

Fixed workflow loading from PNG images to prioritize workflow metadata over parameters, preventing large workflows from being replaced with basic A1111 format.

## Changes

- **What**: Reordered `handleFile()` to check workflow before parameters, added JSON parse error handling and validation, fixed `loadGraphData()` to use explicit type checking instead of falsy check
- **Dependencies**: None

## Review Focus

The key issue was in `handleFile()` where parameters were checked before workflow, causing ComfyUI-generated PNGs (which contain both workflow and parameters metadata) to incorrectly import as A1111 format. The fix ensures:
1. Workflow is always checked first and validated properly
2. Parameters are only used as a fallback when no workflow exists
3. Invalid/malformed workflow data doesn't crash the import process

Additionally, `loadGraphData()` was using a falsy check (`if (!graphData)`) which could incorrectly replace valid but falsy values. Now uses explicit type validation.

Tested with real-world PNG containing 634-node workflow (780KB) + parameters (1KB) - now correctly loads the workflow instead of discarding it.

<!-- Fixes #ISSUE_NUMBER -->

## Screenshots (if applicable)

N/A - Backend logic fix, no UI changes

Fixes #6633

┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-7154-Fix-workflow-loading-from-PNG-images-with-both-workflow-and-parameter-2bf6d73d365081ecb7a6c4bf6b6ccd51) by [Unito](https://www.unito.io)

r-vage committed:
Fix workflow loading from PNG images with both workflow and parameters metadata

- Reorder handleFile() to check workflow before parameters
- Add validation to prevent JSON parse errors from crashing imports
- Fix loadGraphData() to use explicit type validation instead of falsy check
- Ensures ComfyUI-generated PNGs with both metadata types load the workflow, not parameters

Fixes issue where large workflows (e.g., 634 nodes) were replaced with basic A1111 format when importing PNG files.

DrJKL commented on src/scripts/app.ts:
```diff
@@ -1432,6 +1438,37 @@ export class ComfyApp {
+        openSource
+      })
+      return
+    }
```
I'd probably extract a method that returns `ComfyWorkflowJSON|undefined`, but how about this?

```suggestion
    if (workflow) {
      let workflowObj: ComfyWorkflowJSON
      try {
        workflowObj =
          typeof workflow === 'string' ? JSON.parse(workflow) : workflow
      } catch (err) {
        console.error('Failed to parse workflow:', err)
        this.showErrorOnFileLoad(file)
      }

      // Validate workflow is a proper object
      if (
        typeof workflowObj !== 'object' ||
        Array.isArray(workflowObj)
      ) {
        console.error('Invalid workflow structure')
      } else {
        await this.loadGraphData(workflowObj, true, true, fileName, {
          openSource
        })
        return
      }
    }
```

r-vage committed:
Update src/scripts/app.ts

Co-authored-by: Alexander Brown <DrJKL0424@gmail.com>

r-vage committed:
Update workflow object handling and validation

Refactor workflow object handling to allow undefined type and improve validation checks.

r-vage committed:
Refactor workflow parsing and validation logic

r-vage committed:
Fix workflow loading from PNG images with both workflow and parameters metadata

- Reorder handleFile() to check workflow before parameters
- Add validation to prevent JSON parse errors from crashing imports, fallback to parameters or prompt
- Fix loadGraphData() to use explicit type validation instead of falsy check
- Ensures ComfyUI-generated PNGs with both metadata types load the workflow, not parameters

Fixes issue where large workflows (e.g., 634 nodes) were replaced with basic A1111 format when importing PNG files.

r-vage committed:
Merge branch 'main' of https://github.com/r-vage/ComfyUI_frontend

DrJKL committed:
Merge branch 'main' into main

mirabarukaso commented:
# The parameters section should under prompt

The API workflow uses prompt to stroge workflow because we don't care about node position.

Otherwise, it won't fully fix for API generated.
